code/modules/crafting/craft.dm

```diff
@@ -151,152 +151,121 @@
-							RG = locate(RG.type) in Deletion
-							RG.volume = amt
-							RG.data += data
-							continue main_loop
-						else
-							surroundings -= RC
-							amt -= RG.volume
-							RC.reagents.reagent_list -= RG
-							RC.reagents.conditional_update(RC)
-							RGNT = locate(RG.type) in Deletion
-							RGNT.volume += RG.volume
-							RGNT.data += RG.data
-							qdel(RG)
-						RC.on_reagent_change()
-					else
-						surroundings -= RC
-			else if(ispath(A, /obj/item/stack))
-				var/obj/item/stack/S
-				var/obj/item/stack/SD
-				while(amt > 0)
-					S = locate(A) in surroundings
-					if(S.get_amount() >= amt)
-						if(!locate(S.type) in Deletion)
-							SD = new S.type()
-							Deletion += SD
-						S.use(amt)
-						SD = locate(S.type) in Deletion
-						SD.amount += amt
-						continue main_loop
-					else
-						amt -= S.amount
-						if(!locate(S.type) in Deletion)
-							Deletion += S
-						else
-							data = S.amount
-							S = locate(S.type) in Deletion
-							S.add(data)
-						surroundings -= S
-			else
-				var/atom/movable/I
-				while(amt > 0)
-					I = locate(A) in surroundings
-					Deletion += I
-					surroundings -= I
-					amt--
-	var/list/partlist = list(R.parts.len)
-	for(var/M in R.parts)
-		partlist[M] = R.parts[M]
-	for(var/A in R.parts)
-		if(istype(A, /datum/reagent))
-			var/datum/reagent/RG = locate(A) in Deletion
-			if(RG.volume > partlist[A])
-				RG.volume = partlist[A]
-			. += RG
-			Deletion -= RG
-			continue
-		else if(istype(A, /obj/item/stack))
-			var/obj/item/stack/ST = locate(A) in Deletion
-			if(ST.amount > partlist[A])
-				ST.amount = partlist[A]
-			. += ST
-			Deletion -= ST
-			continue
+/datum/personal_crafting/proc/requirements_deletion(datum/crafting_recipe/recipe, mob/user)
+	var/list/surroundings = get_environment(user)
+	var/list/parts_list_for_return = list()
+	var/list/reagent_containers_for_deletion = list()
+	var/list/item_stacks_for_deletion = list()
+
+	for(var/thing in recipe.reqs)
+		var/needed_amount = recipe.reqs[thing]
+		if(ispath(thing, /datum/reagent))
+			var/datum/reagent/part_reagent = locate(thing) in parts_list_for_return
+			if(!part_reagent)
+				part_reagent = new thing()
+				parts_list_for_return += part_reagent
+
+			for(var/obj/item/reagent_containers/container in (surroundings - reagent_containers_for_deletion))
+				var/datum/reagent/contained_reagent = container.reagents.get_reagent(thing)
+				if(!contained_reagent)
+					continue
+				
+				var/extracted_amount = min(contained_reagent.volume, needed_amount)
+				reagent_containers_for_deletion[container] = list(contained_reagent, extracted_amount)
+				part_reagent.volume += extracted_amount
+				part_reagent.data += contained_reagent.data
+				needed_amount -= extracted_amount
+				if(needed_amount <= 0)
+					break
+	
+			if(needed_amount > 0)
+				stack_trace("Bad things happened in [__FILE__] at [__LINE__], we didn't have as many reagents as we expected!")
```

Comment:
```suggestion
				stack_trace("While crafting [recipe], some of [thing] went missing (still need [needed_amount])!")
				continue // ignore the error, and continue crafting for player's benefit
```
stacktrace already gives the filename and the line, but it *doesn't* give the locals.

Reply:
If I remember correctly `stack_trace` gives the filename and the line where `stack_trace` located, but not the actual place where the runtime error happened

Reply:
```
[2022-05-09T07:06:45] Starting up. Round ID is NULL
[2022-05-09T07:06:45] -------------------------
[2022-05-09T07:06:45] Running ParaCode commit: bdfac98376220733305d7d7af6da04c733c04462 (Date: 2022-05-08)
[2022-05-09T07:08:49] Runtime in unsorted.dm,1932: hello
[2022-05-09T07:08:49]   proc name: stack trace (/datum/proc/stack_trace)
[2022-05-09T07:08:49]   usr: Ellie Beach (moxian) (/mob/living/carbon/human)
[2022-05-09T07:08:49]   usr.loc: The shuttle floor (40,143,2) (/turf/simulated/floor/mineral/titanium/blue)
[2022-05-09T07:08:49]   src: /datum/personal_crafting (/datum/personal_crafting)
[2022-05-09T07:08:49]   call stack:
[2022-05-09T07:08:49]   /datum/personal_crafting (/datum/personal_crafting): stack trace("hello")
[2022-05-09T07:08:49]   /datum/personal_crafting (/datum/personal_crafting): requirements deletion(Security HUD removal (/datum/crafting_recipe/hudsunsecremoval), Ellie Beach (/mob/living/carbon/human))
[2022-05-09T07:08:49]   /datum/personal_crafting (/datum/personal_crafting): construct item(Ellie Beach (/mob/living/carbon/human), Security HUD removal (/datum/crafting_recipe/hudsunsecremoval))
[2022-05-09T07:08:49]   /datum/personal_crafting (/datum/personal_crafting): ui act("make", /list (/list), /datum/tgui (/datum/tgui), /datum/ui_state/not_incapacita... (/datum/ui_state/not_incapacitated_state))
[2022-05-09T07:08:49]   /datum/tgui (/datum/tgui): Topic("src=%5B0x21017550%5D_679&actio...", /list (/list))
[2022-05-09T07:08:49]   Moxian (/client): Topic("src=%5B0x21017550%5D_679&actio...", /list (/list), /datum/tgui (/datum/tgui))
```
While I guess it's true that there's no filename/line number here, there is proc name and it should really be enough.

Reply:
Alright then